tfmigrate/state_replace_provider_action_test.go

```diff
@@ -0,0 +1,54 @@
+package tfmigrate
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"testing"
+
+	"github.com/hashicorp/go-version"
+	"github.com/minamijoyo/tfmigrate/tfexec"
+)
+
+func TestAccStateReplaceProviderAction(t *testing.T) {
+	tfexec.SkipUnlessAcceptanceTestEnabled(t)
+
+	backend := tfexec.GetTestAccBackendS3Config(t.Name())
+
+	source := `
+resource "null_resource" "foo" {}
+`
+
+	workspace := "default"
+	tf := tfexec.SetupTestAccWithApply(t, workspace, backend+source)
+	ctx := context.Background()
+
+	actions := []StateAction{
+		NewStateReplaceProviderAction("registry.terraform.io/hashicorp/null", "registry.tfmigrate.io/hashicorp/null"),
+	}
+
+	m := NewStateMigrator(tf.Dir(), workspace, actions, &MigratorOption{}, false)
+	planErr := m.Plan(ctx)
+	if planErr == nil {
+		t.Fatalf("expected migrator plan error but received: %s", planErr)
+	}
+
+	v, err := tf.Version(ctx)
+	if err != nil {
+		t.Fatalf("unexpected version error: %s", err)
+	}
+
+	constraints, err := version.NewConstraint(fmt.Sprintf(">= %s", tfexec.MinimumTerraformVersionForStateReplaceProvider))
+	if err != nil {
+		t.Fatalf("unexpected version constraint error: %s", err)
+	}
+
+	expected := "Could not load the schema for provider\nregistry.tfmigrate.io/hashicorp/null"
+	if !constraints.Check(v) {
+		expected = fmt.Sprintf("configuration uses Terraform version %s; replace-provider action requires Terraform version %s", v, constraints)
+	}
+
+	if !strings.Contains(planErr.Error(), expected) {
+		t.Fatalf("expected migrator plan error to include: %s; received: %s", expected, err)
+	}
```

Comment:
☝️ There may be a better way to author this test. In this case, because we've replaced the legitimate `registry.terraform.io/hashicorp/local` provider with an illegitimate `registry.tfmigrate.io/hashicorp/local` provider, `plan` is expected to error, but we can verify that it does so with the expected illegitimate provider replacement, in effect validating that `replace-provider` did its job.

Reply:
Hmmm, this test is quite tricky. I'll investigate if there is any way to test for a happy path.

Reply:
I couldn't find the best solution. Let me share possible options:

(1) Use a registry proxy such as https://github.com/jasonwbarnett/terraform-registry-proxy
Pros: A general solution for all providers
Cons: Require additional setup for testing

(2) Use any untrusted fork of hashicorp/null and rename to it
Pros: Easy to use
Cons: Potential supply chain attack risk

(3) Fork hashicorp/null by myself
Pros: Trust to use
Cons: Publishing a forked provider to the Terraform Registry takes a lot of work. I don't have experience in developing providers.

(4) Use a real-world example of renaming a namespace for verified providers such as hashicorp/github to integrations/github
Pros: Trust to use
Cons: Need a valid credential
The data source did not result in a plan diff even with namespace renaming, so a resource is needed. However, we do not want to create a real resource for testing. I tried to import a public repository, but it requires a valid GitHub token even a public one.

(5) Embed a legacy v0.12.x tfstate example as a test fixture and force push to remote in advance
Pros: No additional dependency
Cons: As I mentioned in #144, the test may break in the future because the behavior is undefined when an old state file format is loaded with a new version of terraform state commands.

Reply:
As a reasonable compromise, I am leaning towards option (5) and am trying it out, but when I load a legacy tfstate created in v0.12.31 with Terraform v1.5.5, the terraform init fails.
I'm wondering if I need to run state replace-provider before terraform init. Does it work as intended in your upgrade scenario?

```
a05728d678f2:/work/tmp/pr-145/dir5# cat main.tf
resource "null_resource" "foo" {}
a05728d678f2:/work/tmp/pr-145/dir5# terraform -v
Terraform v1.5.5
on linux_arm64

Your version of Terraform is out of date! The latest version
is 1.5.6. You can update by downloading from https://www.terraform.io/downloads.html

a05728d678f2:/work/tmp/pr-145/dir5# terraform init && terraform apply -auto-approve

a05728d678f2:/work/tmp/pr-145/dir5# terraform state list
null_resource.foo

a05728d678f2:/work/tmp/pr-145/dir5# terraform providers

Providers required by configuration:
.
└── provider[registry.terraform.io/hashicorp/null]

Providers required by state:

    provider[registry.terraform.io/hashicorp/null]
```

```
a05728d678f2:/work/tmp/pr-145/dir5# cat legacy.tfstate
{
  "version": 4,
  "terraform_version": "0.12.31",
  "serial": 1,
  "lineage": "3cb3461c-0eab-a7b2-e2e0-57a93c3cc2a1",
  "outputs": {},
  "resources": [
    {
      "mode": "managed",
      "type": "null_resource",
      "name": "foo",
      "provider": "provider.null",
      "instances": [
        {
          "schema_version": 0,
          "attributes": {
            "id": "4107759474886503934",
            "triggers": null
          }
        }
      ]
    }
  ]
}

a05728d678f2:/work/tmp/pr-145/dir5# terraform state push -force legacy.tfstate

a05728d678f2:/work/tmp/pr-145/dir5# terraform providers

Providers required by configuration:
.
└── provider[registry.terraform.io/hashicorp/null]

Providers required by state:

    provider[registry.terraform.io/-/null]

a05728d678f2:/work/tmp/pr-145/dir5# terraform init

Initializing the backend...
╷
│ Error: Invalid legacy provider address
│
│ This configuration or its associated state refers to the unqualified provider "null".
│
│ You must complete the Terraform 0.13 upgrade process before upgrading to later versions.
╵
```

```
a05728d678f2:/work/tmp/pr-145/dir5# terraform state replace-provider registry.terraform.io/-/null registry.terraform.io/hashicorp/null
Terraform will perform the following actions:

  ~ Updating provider:
    - registry.terraform.io/-/null
    + registry.terraform.io/hashicorp/null

Changing 1 resources:

  null_resource.foo

Do you want to make these changes?
Only 'yes' will be accepted to continue.

Enter a value: yes

Successfully replaced provider for 1 resources.

a05728d678f2:/work/tmp/pr-145/dir5# terraform providers

Providers required by configuration:
.
└── provider[registry.terraform.io/hashicorp/null]

Providers required by state:

    provider[registry.terraform.io/hashicorp/null]

a05728d678f2:/work/tmp/pr-145/dir5# terraform init

Initializing the backend...

Initializing provider plugins...
- Reusing previous version of hashicorp/null from the dependency lock file
- Using previously-installed hashicorp/null v3.2.1

Terraform has been successfully initialized!

You may now begin working with Terraform. Try running "terraform plan" to see
any changes that are required for your infrastructure. All Terraform commands
should now work.

If you ever set or change modules or backend configuration for Terraform,
rerun this command to reinitialize your working directory. If you forget, other
commands will detect it and remind you to do so if necessary.

a05728d678f2:/work/tmp/pr-145/dir5# terraform plan
null_resource.foo: Refreshing state... [id=4107759474886503934]

No changes. Your infrastructure matches the configuration.

Terraform has compared your real infrastructure against your configuration and found no differences,
so no changes are needed.
```

Reply:
> I'm wondering if I need to run state replace-provider before terraform init. Does it work as intended in your upgrade scenario?

I don't believe you can run `terraform state replace-provider` before first running `terraform init`. Otherwise, you'll run into `Error loading the state` error:

```
terraform state replace-provider registry.terraform.io/-/null registry.terraform.io/hashicorp/null
Error loading the state: Backend initialization required, please run "terraform init": Reason: Initial configuration of th
e requested backend "s3"
```

In my experience, when using a Terraform version >= v0.13.0 to run `terraform state replace-provider` against a legacy v0.12.31 tfstate, it's necessary to...

1. First run `terraform init`; this will fail with the `Error: Invalid legacy provider address` error you mention above. Despite the failure, it's still necessary.
2. Run `terraform state replace-provider`

Reply:
So, to support state replace-provider -/null hashicorp/null for legacy v0.12.x tfstate described in #144, we need a special error handling in the work directory initialization step, right?
I suggest the acceptance test confirm this scenario by embedding the legacy v0.12.x tfstate as a test fixture.

Reply:
Shouldn't we need to fix OverrideBackendToLocal as well as setupWorkDir? The error occurs during rollback but does not throw the error. It works in unintended balance, but it is not so user-friendly.

```
6d3015a26631:/work/tmp/pr-145/dir11# cat test.hcl
migration "state" "test" {
  dir = "."
  actions = [
    "replace-provider registry.terraform.io/-/null registry.terraform.io/hashicorp/null",
  ]
}

6d3015a26631:/work/tmp/pr-145/dir11# tfmigrate plan test.hcl
2023/08/28 14:52:35 [INFO] [runner] load migration file: test.hcl
2023/08/28 14:52:35 [INFO] [migrator] start state migrator plan
2023/08/28 14:52:35 [INFO] [migrator@.] terraform version: 1.5.5
2023/08/28 14:52:35 [INFO] [migrator@.] initialize work dir
2023/08/28 14:52:35 [INFO] [migrator@.] ignoring error 'Error: Invalid legacy provider address' initilizing work dir; the error is expected when using Terraform >= 0.13 with a legacy Terraform state
2023/08/28 14:52:35 [INFO] [migrator@.] get the current remote state
2023/08/28 14:52:35 [INFO] [migrator@.] override backend to local
2023/08/28 14:52:35 [INFO] [executor@.] create an override file
2023/08/28 14:52:35 [INFO] [migrator@.] creating local workspace folder in: terraform.tfstate.d/default
2023/08/28 14:52:35 [INFO] [executor@.] switch backend to local
2023/08/28 14:52:35 [INFO] [migrator@.] compute a new state
2023/08/28 14:52:36 [INFO] [migrator@.] check diffs
2023/08/28 14:52:36 [INFO] [executor@.] remove the override file
2023/08/28 14:52:36 [INFO] [executor@.] remove the workspace state folder
2023/08/28 14:52:36 [INFO] [executor@.] switch back to remote
2023/08/28 14:52:36 [ERROR] [executor@.] failed to switch back to remote: failed to run command (exited 1): terraform init -input=false -no-color -reconfigure
stdout:

Initializing the backend...

Successfully configured the backend "s3"! Terraform will automatically
use this backend unless the backend configuration changes.

stderr:

Error: Invalid legacy provider address

This configuration or its associated state refers to the unqualified provider
"null".

You must complete the Terraform 0.13 upgrade process before upgrading to
later versions.


2023/08/28 14:52:36 [ERROR] [executor@.] please re-run terraform init -reconfigure
2023/08/28 14:52:36 [INFO] [migrator] state migrator plan success!

6d3015a26631:/work/tmp/pr-145/dir11# echo $?
0
```

Reply:
My intuition was that the error would also occur when switching the backend to local, but I wonder why it doesn't.
Switching to local with a direct aws s3 cp of a v0.12.31 tfstate doesn't cause an error, but with terraform v1.5.5, After state push, switching to local also gives an error.
Rollback to remote will give an error in either case.
I don't know why 🤷‍♂️

Reply:
Anyway, it is better to fix the OverrideBackendToLocal function to suppress the legacy state error in both the switch and rollback phases, isn't it?

Reply:
👍 Good call! I took a stab at implementing your suggestion in https://github.com/minamijoyo/tfmigrate/pull/145/commits/37e28346ecda3b006e4f69dcc666b8f666814eef. How's that look?